app/objects/impersonation/authenticator.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Impersonation
+  class Authenticator
+    def initialize(encrypted_data)
+      @encrypted_data = encrypted_data
+    end
+
+    def build_auth_headers!
+      user.tokens.filter! { |_token, attrs| attrs['impersonated_by'] != admin_user_id }
+      user.build_auth_headers(token.token, token.client) if user.save!
```

Comment:
`user.save!` needs to be called after `user.create_token...` right?

Reply:
I've tested this locally and it seems like it's not necessary.

Reply:
hmm strange, is there some code saving the user after the create_token method is called then?